packages/indexer/indexer/main.py

```diff
@@ -0,0 +1,131 @@
+import asyncio, aiohttp, json, asyncpg, os
+from typing import List
+from dotenv import load_dotenv
+from indexer.chain_mapper import CosmosChain, chain_mapping
+from indexer.data import get_block, get_txs
+from indexer.db import create_tables, drop_tables, get_missing_blocks, upsert_block
+import traceback
+
+load_dotenv()
+
+
+async def process_block(
+    chain: CosmosChain,
+    height: int,
+    pool,
+    session: aiohttp.ClientSession,
+    block_data: dict = None,
+) -> bool:
+    if block_data is None:
+        # this is because block data might be passed in from the live chain data (so removing a rerequest)
+        block_data = await get_block(
+            session, chain.apis[chain.current_api_index], height
+        )
+
+    # need to come back to this
+    txs_data = await get_txs(session, chain.apis[chain.current_api_index], height)
+    if txs_data is None or block_data is None:
+        print(
+            f"{'tx_data' if txs_data is None else 'block_data'} is None - {chain.chain_id}"
+        )
+        return False
+    try:
+        await upsert_block(pool, block_data, txs_data)
+        return True
+    except Exception as e:
+        print(f"upsert_block error {repr(e)} - {chain.chain_id} - {height}")
+        traceback.print_exc()
+        return False
+
+
+async def get_live_chain_data(
+    chain: CosmosChain,
+    pool,
+    session: aiohttp.ClientSession,
+):
+    last_block = 0
+    while True:
+        try:
+            block_data = await get_block(session, chain.apis[chain.current_api_index])
+            if block_data is not None:
+                current_block = int(block_data["block"]["header"]["height"])
+                if last_block >= current_block:
+                    # print(f"{chain_id} - block {current_block} already indexed")
+                    await asyncio.sleep(1)
+                else:
+                    last_block = current_block
+                    if await process_block(
+                        chain=chain,
+                        height=current_block,
+                        pool=pool,
+                        session=session,
+                        block_data=block_data,
+                    ):
+                        print(
+                            f"processed live block {current_block} - {chain.chain_id}"
+                        )
+                    else:
+                        # should we save this to db?
```

Comment:
We could have an error table but I'm afraid of things continually just piling up in there. With GCP we can enable logging and just search those logs. There are some tools to make this easier but let's ignore for now.

Reply:
it seems dumb to keep  reproccess blocks that keep causing errors imo

Reply:
Yeah agreed.